src/main/webapp/script.js

```diff
@@ -69,37 +79,132 @@ function toggleAQLayer() {
 }
 
 function populateAQVisualisationData() {
-    fetch("/visualisation").then(response => response.json()).then((data) => {
-        const aqData = [];
-        for (i = 0; i < data.length; i++) {
-            const dataItem = data[i];
-            let dataPoint = {};
-            dataPoint.location = new google.maps.LatLng(data[i].lat,data[i].long);
-            dataPoint.weight = dataItem.aqi;
-            aqData.push(dataPoint);
-        }
-        aqData.push({location: new google.maps.LatLng(-37.785, 150.435), weight: 3})
+    let fetchURL = "/visualisation";
+    fetchURL += `?zoom-level=${map.getZoom()}`;
+    const mapBounds = map.getBounds();
+    const swCorner = mapBounds.getSouthWest();
+    const neCorner = mapBounds.getNorthEast();
+    fetchURL += `&sw-lat=${swCorner.lat()}`;
+    fetchURL += `&sw-long=${swCorner.lng()}`;
+    fetchURL += `&ne-lat=${neCorner.lat()}`;
+    fetchURL += `&ne-long=${neCorner.lng()}`;
+    fetch(fetchURL).then(response => response.json()).then((data) => {
+        const aqData = convertGriddedDataToWeightedPoints(data);
         loadHeatmap(aqData);
     });
 }
 
 function loadHeatmap(data) {
-    aqLayer = new google.maps.visualization.HeatmapLayer({
-        data: data,
-        gradient: [
-            "rgba(65,169,60,0)",
-            "rgba(65, 169, 60, 1)",
-            "rgba(238, 201, 0, 1)",
-            "rgba(228, 116, 0, 1)",
-            "rgba(186, 0, 41, 1)",
-            "rgba(86, 3, 23, 1)",
-        ],
-        maxIntensity: 200,
-    });
+    if (!aqLayer) {
+        aqLayer = new google.maps.visualization.HeatmapLayer({
+            gradient: [
+                "rgba(65,169,60,0)",
+                "rgba(65, 169, 60, 1)",
+                "rgba(238, 201, 0, 1)",
+                "rgba(228, 116, 0, 1)",
+                "rgba(186, 0, 41, 1)",
+                "rgba(86, 3, 23, 1)",
+            ],
+            maxIntensity: 200,
+            dissipating: true,
+            radius: 100, // arbitrary value used to make the points easily visible
+        });
+    }
+    aqLayer.setData([]);
```

Comment:
Why set it to `[]` first?

Reply:
To remove the old data so that duplicates don't occur

Reply:
Why would `aqLayer.setData([]);` remove data points, and `aqLayer.setData(data);` not remove data points? Does the API treat an empty list specially?

Reply:
It appears to work without it so I've removed that line, I think I had both initially because there was a stackoverflow that said to set it to empty.